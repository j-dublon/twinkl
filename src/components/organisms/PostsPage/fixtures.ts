import { Mock } from "vitest";

export const mockDeletePost: Mock = vi.fn();
