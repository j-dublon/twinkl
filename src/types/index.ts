export type Post = {
  id: number;
  title: string;
  body: string;
  userId: number;
};

export type PostPageProps = {
  posts: Post[];
  removePost: (id: number) => void;
};

export type PostListProps = {
  posts: Post[];
  removePost: (id: number) => void;
};

export type PostCardProps = {
  post: Post;
  removePost: (id: number) => void;
};
