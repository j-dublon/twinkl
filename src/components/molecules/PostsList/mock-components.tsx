import { PostCardProps } from "@/types";
import { FC } from "react";

export const MockPostCard: FC<PostCardProps> = ({ post }) => (
  <div>{post.title}</div>
);
