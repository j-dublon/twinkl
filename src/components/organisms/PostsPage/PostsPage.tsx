import { PostPageProps } from "@/types";
import { FC } from "react";

export const PostsPage: FC<PostPageProps> = ({ posts }) => {
  return (
    <>
      {posts.map((post) => (
        <div>{post.title}</div>
      ))}
    </>
  );
};
