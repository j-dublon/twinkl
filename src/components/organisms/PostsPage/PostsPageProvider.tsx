import { deletePost, fetchAllPosts } from "@/services/posts";
import { Post } from "@/types";
import { FC, useEffect, useState } from "react";
import { PostsPage } from "./PostsPage";
import { Error, Loading } from "../../atoms";

export const PostsPageProvider: FC = () => {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);
  const [posts, setPosts] = useState<Post[]>([]);

  useEffect(() => {
    const getPosts = async () => {
      setLoading(true);
      const posts = await fetchAllPosts();
      if (!posts) {
        setError(true);
      } else {
        setPosts(posts);
      }
      setLoading(false);
    };
    getPosts();
  }, []);

  const handleRemovePost = async (postId: number) => {
    setLoading(true);
    const res = await deletePost(postId);
    if (res.status !== 200) {
      // Show user a toast message
      // E.g. Sorry, there was a problem deleting the post, please try again
    }
    setLoading(false);
  };

  return (
    <>
      {loading ? (
        <Loading />
      ) : error ? (
        <Error />
      ) : (
        <PostsPage posts={posts} removePost={handleRemovePost} />
      )}
    </>
  );
};
