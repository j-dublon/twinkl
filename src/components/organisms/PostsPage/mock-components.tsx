import { PostListProps } from "@/types";
import { FC } from "react";

export const MockPostsList: FC<PostListProps> = () => (
  <div>Mock Posts List</div>
);

export const MockEmptyList: FC = () => <div>Mock Empty List</div>;
