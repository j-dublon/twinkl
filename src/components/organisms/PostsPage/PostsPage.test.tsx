import { vi } from "vitest";
import { render } from "@testing-library/react";
import { PostsPage } from "./PostsPage";
import { mockPosts } from "@/tests/mockData";
import { MockEmptyList, MockPostsList } from "./mock-components";
import { PostListProps } from "@/types";

vi.mock("../../molecules/PostsList/PostsList", () => ({
  PostsList: (props: PostListProps) => <MockPostsList {...props} />,
}));

vi.mock("../../atoms/EmptyList/EmptyList", () => ({
  EmptyList: () => <MockEmptyList />,
}));

describe("Component: PostsPage", () => {
  it("SHOULD match snapshot", () => {
    const component = render(<PostsPage posts={mockPosts} />);
    expect(component).toMatchSnapshot();
  });

  it("SHOULD render the PostsList component WHEN posts are available", () => {
    const { getByText } = render(<PostsPage posts={mockPosts} />);
    expect(getByText("Mock Posts List"));
  });

  it("SHOULD render the EmptyList component WHEN posts are not available", () => {
    const { getByText } = render(<PostsPage posts={[]} />);
    expect(getByText("Mock Empty List"));
  });
});
