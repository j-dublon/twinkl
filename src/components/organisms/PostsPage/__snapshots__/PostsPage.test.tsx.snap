// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component: PostsPage > SHOULD match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-label="Post content"
        class="m-4 pb-1 focus:outline-none focus:ring-2 focus:ring-blue-500 border-b-1 border-b-zinc-300"
        tabindex="0"
      >
        <h1>
          sunt aut facere repellat provident occaecati excepturi optio reprehenderit
        </h1>
        <p>
          quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
        </p>
      </div>
      <div
        aria-label="Post content"
        class="m-4 pb-1 focus:outline-none focus:ring-2 focus:ring-blue-500 border-b-1 border-b-zinc-300"
        tabindex="0"
      >
        <h1>
          qui est esse
        </h1>
        <p>
          est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla
        </p>
      </div>
      <div
        aria-label="Post content"
        class="m-4 pb-1 focus:outline-none focus:ring-2 focus:ring-blue-500 border-b-1 border-b-zinc-300"
        tabindex="0"
      >
        <h1>
          ea molestias quasi exercitationem repellat qui ipsa sit aut
        </h1>
        <p>
          et iusto sed quo iure
voluptatem occaecati omnis eligendi aut ad
voluptatem doloribus vel accusantium quis pariatur
molestiae porro eius odio et labore et velit aut
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-label="Post content"
      class="m-4 pb-1 focus:outline-none focus:ring-2 focus:ring-blue-500 border-b-1 border-b-zinc-300"
      tabindex="0"
    >
      <h1>
        sunt aut facere repellat provident occaecati excepturi optio reprehenderit
      </h1>
      <p>
        quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
      </p>
    </div>
    <div
      aria-label="Post content"
      class="m-4 pb-1 focus:outline-none focus:ring-2 focus:ring-blue-500 border-b-1 border-b-zinc-300"
      tabindex="0"
    >
      <h1>
        qui est esse
      </h1>
      <p>
        est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla
      </p>
    </div>
    <div
      aria-label="Post content"
      class="m-4 pb-1 focus:outline-none focus:ring-2 focus:ring-blue-500 border-b-1 border-b-zinc-300"
      tabindex="0"
    >
      <h1>
        ea molestias quasi exercitationem repellat qui ipsa sit aut
      </h1>
      <p>
        et iusto sed quo iure
voluptatem occaecati omnis eligendi aut ad
voluptatem doloribus vel accusantium quis pariatur
molestiae porro eius odio et labore et velit aut
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
